packages/backend/src/index.ts

```diff
@@ -55,13 +55,20 @@ import graphql from './plugins/graphql';
 import app from './plugins/app';
 import badges from './plugins/badges';
 import jenkins from './plugins/jenkins';
+import permission from './plugins/permission';
 import { PluginEnvironment } from './types';
+import { ServerPermissionClient } from '@backstage/plugin-permission-node';
 
 function makeCreateEnv(config: Config) {
   const root = getRootLogger();
   const reader = UrlReaders.default({ logger: root, config });
   const discovery = SingleHostDiscovery.fromConfig(config);
-  const tokenManager = ServerTokenManager.noop();
+  const tokenManager = ServerTokenManager.fromConfig(config);
```

Comment:
If the app is configured with the `noop` token manager, the permission client can't authenticate backend to backend requests (see `ServerPermissionClient` below), so we need `ServerTokenManager` to be configured with a real secret. @mtlewis and I were discussing how it might be nice to generate these each time you start the backend the way it's done for sign in when you don't have a database set up. This would be a much better experience than not being able to start the example backend at all if you don't set up a backend to backend secret. Thoughts?

Reply:
I'm thinking we could just hardcode the secret to `'INSECURE_DEVELOPMENT_SECRET'` or something like that?

Reply:
Hmm - in app-config.yaml? If we do that, it feels a bit too easy for people to ship insecure production deployments, no? Feels like a solution with generated secrets has the nice property that it will work for local development but break if you deploy it (with multiple instances).

Reply:
Yeah it'd make sense too, realizing that people are ofc going to copy our setup. What do you think might be a good API for that? Thinking we want it out of the way but still avoid confusion where people deploy with generated secrets by mistake

Reply:
Wondering if we can leverage some of the same things that are used in auth-backend for generating secrets are runtime (MemoryKeystore, DatabaseKeystore etc). Seems like a similar pattern could be viable for backend secrets too, but I'm very new to that part of the code so I could definitely be missing some nuance!

Reply:
The auth-backend solution has the advantage of only needing to share the database with other auth-backend instances. In this case we'd be needing to share a database between all backend instances, which I don't think is something that would be good to encourage. I agree it's nice to keep things configuration-free when possible though, so all up for searching for a solution for a while here.

I'm not expecting to find a configuration free solution for a production deployment here really though. How would a new instance join if they can't already prove that they are trusted, which is the problem we're already trying to solve :grin:

What I'm thinking for now is just something based on generated keys that works well for development, but would just fail to start up in production. For example if `NODE_ENV !== 'development'` we could reject a generated secret.

Reply:
>  In this case we'd be needing to share a database between all backend instances, which I don't think is something that would be good to encourage.

Right - makes sense, thanks for pointing it out.

> What I'm thinking for now is just something based on generated keys that works well for development, but would just fail to start up in production. For example if NODE_ENV !== 'development' we could reject a generated secret.

Like that option! Perhaps we should print a warning in the console when we generate a secret, to warn users that they'll need to configure one before deploying to prod? 

The biggest issue for me today is that using the noop ServerTokenManager causes the authorize method in the PermissionClient to return early, which means the whole authorization framework is sidestepped in what is really quite a hard-to-discover way. I like the fact that changing over to using `ServerTokenManager.fromConfig` in the example backend means that the setup will work locally without any configuration changes.

Reply:
One other alternative: I wonder if we should reconsider changing the "no secret" token manager. We could replace `ServerTokenManager.noop` (that returns empty string tokens and validates anything) with `ServerTokenManager.insecure`, which generates and validates identifiable but unsigned JWTs. I think this removes the strict coupling between backend-to-backend auth and the permissions framework.

Reply:
Just pushed up the changes that generate a secret if in development mode. PTAL!

Reply:
Nice! Looking at the code again, I want to flag that this change means that generating and providing a secret will be mandatory for all integrators using the DefaultCatalogCollator or the DefaultTechDocsCollator search collators (which I imagine covers almost everyone). Both of these require a token manager, and after this PR, there's no way of constructing one in prod without a secret in config.

Reply:
>  I want to flag that this change means that generating and providing a secret will be mandatory for all integrators using the DefaultCatalogCollator or the DefaultTechDocsCollator search collators (which I imagine covers almost everyone).

I'm not familiar with the collators but it sounds like there are some docs that we should update on the collators' end to point to the backend-to-backend docs. Are there other steps we need to take that I'm missing?

Reply:
My point is not really about collators - it's more to highlight the fact that since almost everyone uses them, this change means that we're now forcing almost everyone to generate and configure a secret, where before people could choose to ignore backend-to-backend authentication by using a noop ServerTokenManager. cc @Rugvip for your thoughts too!

Reply:
Just had a quick catch up with @rugvip - we came up with the following proposal for a few additional changes on top of this PR

1. Reintroduce the noop server token manager to be used in create-app as the default token manager for consumers like the collators. We'll keep both the development-only secret generating logic in fromConfig, and the noop token manager.
2. The ServerPermissionClient currently checks whether it has a valid server token, _then_ whether the permission system is enabled at all, _then_ calls the permission-backend. This order is convenient because it's extending PermissionClient, but it's probably not the right order. We should adjust so that it first checks whether the permission system is enabled, _then_ validates the server token. That way when permissions are disabled, the noop token manager can be used without fouling up the logic inside the ServerPermissionClient, because the code path won't be reached.
3. Inside ServerPermissionClient, throw an error if permissions are enabled and the token manager is a noop token manager. We'll need some kind of mechanism to differentiate between a secure ServerTokenManager with a valid configured secret and the noop one. We can consider doing this in the constructor or the authorize call. In either case, we should return a descriptive error message, that makes it clear to the user that it's mandatory to configure a backend secret when the permissions system is enabled.

Reply:
Just pushed up [a commit with these changes](https://github.com/backstage/backstage/pull/8357/commits/5732dadfa2fafa85eceacda69f1ba5efe5d1f516).